### PR TITLE
[core] IndexLayer memory layout introspection

### PR DESCRIPTION
**Pull request by MathiasPaulin**
This PR adds access to the indexLayer of a mesh with small introspection capabilities.

This allows to use any indexLayer transparently when using low level  rendering API (e.g. OpenGL, Vulkan, embree, ...).
The way one can introspect the indexLayer is the same than any mesh attribute.

### CODE REVIEW

**Review by nmellado — APPROVED** [review]

**Comment by dlyr** [comment]
As a review I pushed some modification.

File: src/Core/Utils/Attribs.cpp [hunk 1]
```diff
@@ -11,12 +11,12 @@ AttribBase::~AttribBase() {
 }
 
 template <>
-size_t Attrib<float>::getElementSize() const {
+size_t Attrib<float>::getNumberOfComponent() const {
```
Comment (nmellado) [inline src/Core/Utils/Attribs.cpp:14]:
```suggestion
size_t Attrib<float>::getNumberOfComponents() const {
```

File: src/Core/Utils/Attribs.cpp [hunk 2]
```diff
@@ -11,12 +11,12 @@ AttribBase::~AttribBase() {
...
     return 1;
 }
 
 template <>
-size_t Attrib<double>::getElementSize() const {
+size_t Attrib<double>::getNumberOfComponent() const {
```
Comment (nmellado) [inline src/Core/Utils/Attribs.cpp:19]:
```suggestion
size_t Attrib<double>::getNumberOfComponents() const {
```

File: src/Core/Geometry/IndexedGeometry.inl [hunk 3]
```diff
@@ -65,10 +65,30 @@ inline bool GeometryIndexLayer<T>::operator==( const GeometryIndexLayerBase& oth
 }
 
 template <typename T>
-inline size_t GeometryIndexLayer<T>::size() {
+inline size_t GeometryIndexLayer<T>::getSize() const {
     return m_collection.size();
 }
 
+template <typename T>
+inline size_t GeometryIndexLayer<T>::getNumberOfComponent() const {
```
Comment (nmellado) [inline src/Core/Geometry/IndexedGeometry.inl:73]:
```suggestion
inline size_t GeometryIndexLayer<T>::getNumberOfComponents() const {
```

File: src/Core/Geometry/IndexedGeometry.hpp [hunk 4]
```diff
@@ -63,9 +63,19 @@ struct GeometryIndexLayer : public GeometryIndexLayerBase {
     /// \warning Does not account for elements permutations
     inline bool operator==( const GeometryIndexLayerBase& other ) const final;
 
-    inline size_t size() override;
+    inline size_t getSize() const override final;
 
-    inline GeometryIndexLayerBase* clone() override;
+    inline GeometryIndexLayerBase* clone() override final;
+
+    inline size_t getNumberOfComponent() const override final;
```
Comment (nmellado) [inline src/Core/Geometry/IndexedGeometry.hpp:70]:
```suggestion
    inline size_t getNumberOfComponents() const override final;
```

File: src/Core/Utils/Attribs.hpp [hunk 5]
```diff
@@ -136,7 +115,7 @@ class Attrib : public AttribBase
 
     size_t getSize() const override;
 
-    size_t getElementSize() const override;
+    size_t getNumberOfComponent() const override;
```
Comment (nmellado) [inline src/Core/Utils/Attribs.hpp:118]:
```suggestion
    size_t getNumberOfComponents() const override;
```

File: src/Core/Utils/Attribs.inl [hunk 6]
```diff
@@ -131,12 +131,12 @@ bool Attrib<T>::isType() {
 // user wants to have a specific attrib precision, while Radium Scalar is the other precision.
 // Do we need to specialize for all floating point types ?
 template <>
-RA_CORE_API size_t Attrib<float>::getElementSize() const;
+RA_CORE_API size_t Attrib<float>::getNumberOfComponent() const;
```
Comment (nmellado) [inline src/Core/Utils/Attribs.inl:134]:
```suggestion
RA_CORE_API size_t Attrib<float>::getNumberOfComponents() const;
```

File: src/Core/Utils/Attribs.inl [hunk 7]
```diff
@@ -131,12 +131,12 @@ bool Attrib<T>::isType() {
...
 template <>
-RA_CORE_API size_t Attrib<double>::getElementSize() const;
+RA_CORE_API size_t Attrib<double>::getNumberOfComponent() const;
```
Comment (nmellado) [inline src/Core/Utils/Attribs.inl:136]:
```suggestion
RA_CORE_API size_t Attrib<double>::getNumberOfComponents() const;
```

File: src/Core/Utils/Attribs.inl [hunk 8]
```diff
@@ -131,12 +131,12 @@ bool Attrib<T>::isType() {
...
 // template specialization defined in header.
 template <typename T>
-size_t Attrib<T>::getElementSize() const {
+size_t Attrib<T>::getNumberOfComponent() const {
```
Comment (nmellado) [inline src/Core/Utils/Attribs.inl:139]:
```suggestion
size_t Attrib<T>::getNumberOfComponents() const {
```

File: src/Core/Utils/ContainerIntrospectionInterface.hpp [hunk 9]
```diff
@@ -0,0 +1,35 @@
+#pragma once
+
+namespace Ra {
+namespace Core {
+namespace Utils {
+
```
Comment (nmellado) [inline src/Core/Utils/ContainerIntrospectionInterface.hpp:6]: Could you add some comments ? a brief at least ?

File: src/Core/Utils/ContainerIntrospectionInterface.hpp [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+  public:
+    virtual ~ContainerIntrospectionInterface() {}
+    /// \return the number of element (i.e. faces, vector) contained in the array.
+    virtual size_t getSize() const = 0;
+
+    /// Return the number of components of one element (i.e. for
+    /// glVertexAttribPointer size value)
+    /// \see https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/glVertexAttribPointer.xhtml
+    /// \return the number of components of one element (i.e. faces, vector)
+    virtual size_t getNumberOfComponent() const = 0;
```
Comment (nmellado) [inline src/Core/Utils/ContainerIntrospectionInterface.hpp:18]:
```suggestion
    virtual size_t getNumberOfComponents() const = 0;
```

File: src/Core/Utils/ContainerIntrospectionInterface.hpp (hunk 9 above)
Comment (nmellado) [reply src/Core/Utils/ContainerIntrospectionInterface.hpp:6]: Question: is this introspection interface motivated (and valid) only for OpenGL bindings ?

File: src/Engine/Data/Mesh.inl [hunk 11]
```diff
@@ -136,13 +136,13 @@ void IndexedAttribArrayDisplayable<I>::autoVertexAttribPointer( const ShaderProg
 #ifdef CORE_USE_DOUBLE
             binding->setBuffer( m_vbos[m_handleToBuffer[attribName]].get(),
                                 0,
-                                attrib->getElementSize() * sizeof( float ) );
+                                attrib->getNumberOfComponent() * sizeof( float ) );
```
Comment (nmellado) [inline src/Engine/Data/Mesh.inl:139]:
```suggestion
                                attrib->getNumberOfComponents() * sizeof( float ) );
```

File: src/Engine/Data/Mesh.inl [hunk 12]
```diff
@@ -136,13 +136,13 @@ void IndexedAttribArrayDisplayable<I>::autoVertexAttribPointer( const ShaderProg
...
 #else
 
             binding->setBuffer(
                 m_vbos[m_handleToBuffer[attribName]].get(), 0, attrib->getStride() );
 #endif
-            binding->setFormat( attrib->getElementSize(), GL_SCALAR );
+            binding->setFormat( attrib->getNumberOfComponent(), GL_SCALAR );
```
Comment (nmellado) [inline src/Engine/Data/Mesh.inl:145]:
```suggestion
            binding->setFormat( attrib->getNumberOfComponents(), GL_SCALAR );
```

File: src/Engine/Data/Mesh.inl [hunk 13]
```diff
@@ -264,13 +264,13 @@ void CoreGeometryDisplayable<CoreGeometry>::autoVertexAttribPointer( const Shade
 #ifdef CORE_USE_DOUBLE
             binding->setBuffer( m_vbos[m_handleToBuffer[attribName]].get(),
                                 0,
-                                attrib->getElementSize() * sizeof( float ) );
+                                attrib->getNumberOfComponent() * sizeof( float ) );
```
Comment (nmellado) [inline src/Engine/Data/Mesh.inl:267]:
```suggestion
                                attrib->getNumberOfComponents() * sizeof( float ) );
```

File: src/Engine/Data/Mesh.inl [hunk 14]
```diff
@@ -264,13 +264,13 @@ void CoreGeometryDisplayable<CoreGeometry>::autoVertexAttribPointer( const Shade
...
 #else
 
             binding->setBuffer(
                 m_vbos[m_handleToBuffer[attribName]].get(), 0, attrib->getStride() );
 #endif
-            binding->setFormat( attrib->getElementSize(), GL_SCALAR );
+            binding->setFormat( attrib->getNumberOfComponent(), GL_SCALAR );
```
Comment (nmellado) [inline src/Engine/Data/Mesh.inl:273]:
```suggestion
            binding->setFormat( attrib->getNumberOfComponents(), GL_SCALAR );
```

File: src/Engine/Data/Mesh.inl [hunk 15]
```diff
@@ -347,7 +347,7 @@ void CoreGeometryDisplayable<CoreGeometry>::updateGL() {
                 if ( !m_vbos[idx] ) { m_vbos[idx] = globjects::Buffer::create(); }
 
                 auto stride      = b->getStride();
-                auto eltSize     = b->getElementSize();
+                auto eltSize     = b->getNumberOfComponent();
```
Comment (nmellado) [inline src/Engine/Data/Mesh.inl:350]:
```suggestion
                auto eltSize     = b->getNumberOfComponents();
```

File: tests/unittest/Core/indexview.cpp [hunk 16]
```diff
@@ -208,16 +208,16 @@ TEST_CASE( "Core/Geometry/IndexedGeometry/Attributes", "[Core][Core/Geometry][In
     auto& attribM2_1 = m2.getAttrib( handle1 );
     auto& attribM2_2 = m2.getAttrib( handle2 );
     REQUIRE( attribM2_1.getSize() == 3 );
-    REQUIRE( attribM2_1.getElementSize() == 3 );
+    REQUIRE( attribM2_1.getNumberOfComponent() == 3 );
```
Comment (nmellado) [inline tests/unittest/Core/indexview.cpp:211]:
```suggestion
    REQUIRE( attribM2_1.getNumberOfComponents() == 3 );
```

File: tests/unittest/Core/indexview.cpp [hunk 17]
```diff
@@ -208,16 +208,16 @@ TEST_CASE( "Core/Geometry/IndexedGeometry/Attributes", "[Core][Core/Geometry][In
...
     REQUIRE( attribM2_1.getStride() == sizeof( Vector3 ) );
     REQUIRE( attribM2_1.getBufferSize() == 3 * sizeof( Vector3 ) );
     attribM2_1.resize( 10 );
     REQUIRE( attribM2_1.getSize() == 10 );
-    REQUIRE( attribM2_1.getElementSize() == 3 );
+    REQUIRE( attribM2_1.getNumberOfComponent() == 3 );
```
Comment (nmellado) [inline tests/unittest/Core/indexview.cpp:216]:
```suggestion
    REQUIRE( attribM2_1.getNumberOfComponents() == 3 );
```

File: tests/unittest/Core/indexview.cpp [hunk 18]
```diff
@@ -208,16 +208,16 @@ TEST_CASE( "Core/Geometry/IndexedGeometry/Attributes", "[Core][Core/Geometry][In
...
     REQUIRE( attribM2_1.getStride() == sizeof( Vector3 ) );
     REQUIRE( attribM2_1.getBufferSize() == 10 * sizeof( Vector3 ) );
     REQUIRE( attribM2_2.getSize() == 3 );
-    REQUIRE( attribM2_2.getElementSize() == 1 );
+    REQUIRE( attribM2_2.getNumberOfComponent() == 1 );
```
Comment (nmellado) [inline tests/unittest/Core/indexview.cpp:220]:
```suggestion
    REQUIRE( attribM2_2.getNumberOfComponents() == 1 );
```

**Comment by nmellado** [comment]
@dlyr thanks, even better that way.

**Comment by dlyr** [comment]
Thanks for the "s", I planned to do it after first refactor, but forgot ;)
Evrything should be fixed now.

**Review by nmellado — APPROVED** [review]

**Review by dlyr — APPROVED** [review]

File: src/Core/Utils/ContainerIntrospectionInterface.hpp (hunk 9 above)
Comment (MathiasPaulin) [reply src/Core/Utils/ContainerIntrospectionInterface.hpp:6]: My comment is perhaps obsolete but this introspection is not limited to OpenGL/GPU buffers but is motivated and interesting when interfacing Radium with any API that takes (eventually shared) data buffers as input (e.g. intel embree is not GPU based but requires this to define data buffers from Radium geometry).

Comment (nmellado) [reply src/Core/Utils/ContainerIntrospectionInterface.hpp:6]: ok thx

**Pull request merged**